jmclient/jmclient/cryptoengine.py

```diff
@@ -278,62 +263,59 @@ def VBYTE(cls):
-        wpkscript = pubkey_to_p2wpkh_script(pubkey)
-        pkscript = pubkey_to_p2pkh_script(pubkey)
-
-        signing_tx = btc.segwit_signature_form(tx, index, pkscript, amount,
-                                               hashcode=hashcode,
-                                               decoder_func=lambda x: x)
-        # FIXME: encoding mess
-        sig = unhexlify(btc.ecdsa_tx_sign(signing_tx, hexlify(privkey).decode('ascii'),
-                                          hashcode=hashcode, **kwargs))
-
-        assert len(wpkscript) == 0x16
-        tx['ins'][index]['script'] = b'\x16' + wpkscript
-        tx['ins'][index]['txinwitness'] = [sig, pubkey]
-
-        return tx
-
+        return btc.sign(btc.serialize(tx), index, privkey,
+                        hashcode=hashcode, amount=amount, native=False)
 
 class BTC_P2WPKH(BTCEngine):
+
     @classproperty
     def VBYTE(cls):
-        return btc.BTC_P2SH_VBYTE[get_network()]
+        """Note that vbyte is needed in the native segwit case
+        to decide the value of the 'human readable part' of the
+        bech32 address. If it's 0 or 5 we use 'bc', else we use
+        'tb' for testnet bitcoin; so it doesn't matter if we use
+        the P2PK vbyte or the P2SH one.
+        However, regtest uses 'bcrt' only (and fails on 'tb'),
+        so bitcoin.script_to_address currently uses an artificial
+        value 100 to flag that case.
+        This means that for testing, this value must be explicitly
+        overwritten.
+        """
+        return btc.BTC_P2PK_VBYTE[get_network()]
 
     @classmethod
     def pubkey_to_script(cls, pubkey):
-        return pubkey_to_p2wpkh_script(pubkey)
+        return unhexlify(btc.pubkey_to_p2wpkh_script(pubkey))
+
+    @classmethod
+    def pubkey_to_script_code(cls, pubkey):
+        """ As per BIP143, the scriptCode for the p2wpkh
+        case is "76a914+hash160(pub)+"88ac" as per the
+        scriptPubKey of the p2pkh case.
+        """
+        return pubkey_to_p2pkh_script(pubkey)
 
     @classmethod
     def sign_transaction(cls, tx, index, privkey, amount,
                          hashcode=btc.SIGHASH_ALL, **kwargs):
         assert amount is not None
-        raise NotImplementedError("The following code is completely untested")
-
-        pubkey = cls.privkey_to_pubkey(privkey)
-        script = cls.pubkey_to_script(pubkey)
-
-        signing_tx = btc.segwit_signature_form(tx, index, script, amount,
-                                               hashcode=hashcode,
-                                               decoder_func=lambda x: x)
-        # FIXME: encoding mess
-        sig = unhexlify(btc.ecdsa_tx_sign(signing_tx, hexlify(privkey),
-                                          hashcode=hashcode, **kwargs))
-
-        tx['ins'][index]['script'] = script
-        tx['ins'][index]['txinwitness'] = [sig, pubkey]
-
-        return tx
-
+        return btc.sign(btc.serialize(tx), index, privkey,
```

Comment:
On this one, I definitely agree with your thinking - not good/logical to pass serializations instead of objects. But I think I'll leave this one alone, for now.